### PR TITLE
[feat][Java client]: Support Reader Interceptor

### LINKED ISSUE #14642: [Feature] Support reader interceptor
**Is your feature request related to a problem? Please describe.**
In [PIP 23](https://github.com/apache/pulsar/wiki/PIP-23%3A-Message-Tracing-By-Interceptors), pulsar has supported producer interceptor and consumer interceptor, but we don't have reader interceptor for now.

To support reader interceptor will allow users to do integration with Pulsar client.

**Describe the solution you'd like**
Introduce a new ReaderInterceptor interface to intercept message before read operation

### PR DESCRIPTION
**Pull request by gaoran10**
Fix #14642 

### Motivation

Currently, the Pulsar [PIP-23](https://github.com/apache/pulsar/wiki/PIP-23%3A-Message-Tracing-By-Interceptors) support adds interceptors for producers and consumers to implement message tracing, but it didn't support adding interceptors for readers, this PR wants to support adding interceptors for readers.

Because the Pulsar reader is based on the Pulsar consumer, we could leverage consumer interceptor to implement reader interceptor, use consumer interceptor to wrap reader interceptor, the reader interceptor events should be a subset of the consumer interceptors.

### Modifications

1. Add a new interface `ReaderInterceptor`, used to customize the reader interceptor.
2. Support set reader interceptor by `ReaderBuilder`.
3. Support set auto-update partition configurations by `ReaderBuilder`.

### Verifying this change

Add ReaderInterceptor test. 

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (yes)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below or label this PR directly (if you have committer privilege).

Need to update docs? 

- [x] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [ ] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Review by codelipenghui — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ReaderInterceptorUtil.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ */
+public class ReaderInterceptorUtil {
+
+    public static <T> ConsumerInterceptors<T> convertToConsumerInterceptors(
+            Reader<T> reader, List<ReaderInterceptor<T>> interceptorList) {
+        if (interceptorList == null || interceptorList.isEmpty()) {
+            return null;
+        }
+        List<ConsumerInterceptor<T>> consumerInterceptorList = new ArrayList<>();
+        for (ReaderInterceptor<T> readerInterceptor : interceptorList) {
```
Comment (Technoboy-) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ReaderInterceptorUtil.java:43]: new ArrayList<>(interceptorList.size());  is better?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ReaderBuilderImpl.java [hunk 2]
```diff
@@ -229,4 +230,26 @@ private ReaderBuilderImpl(PulsarClientImpl client, ReaderConfigurationData<T> co
...
+
+    @Override
+    public ReaderBuilder<T> autoUpdatePartitions(boolean autoUpdate) {
+        this.conf.setAutoUpdatePartitions(autoUpdate);
+        return this;
+    }
+
+    @Override
+    public ReaderBuilder<T> autoUpdatePartitionsInterval(int interval, TimeUnit unit) {
+        checkArgument(interval > 0, "interval needs to be > 0");
```
Comment (Technoboy-) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ReaderBuilderImpl.java:242]: Do we need to keep this value at least 1 minute?
If user configs with 1ms, then what happens ?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ReaderInterceptorUtil.java (hunk 1 above)
Comment (gaoran10) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ReaderInterceptorUtil.java:43]: fix

**Review by Technoboy- — APPROVED** [review]

**Review by Demogorgon314 — APPROVED** [review]

**Comment by Anonymitaet** [comment]
@momo-jun a soft reminder: here is a PR targeted at 2.11 w/ `doc-required` label.

**Comment by Anonymitaet** [comment]
@momo-jun a soft reminder: this PR is merged, doc work can be started, thanks.

**Comment by momo-jun** [comment]
> @momo-jun a soft reminder: this PR is merged, doc work can be started, thanks.

@gaoran10 will start to work on a draft as soon as he cleaned up his ongoing tasks.

**Pull request merged**